File: src/main/kotlin/RBTree.kt
enum class Color {
  Red, Black
}

interface Node<K: Comparable<K>>{
  fun isRed(): Boolean
  fun insert(k: K): Branch<K>{
    val br = insert_(k)
    return Branch(Color.Black, br.left, br.key, br.right)
  }
  fun insert_(k: K): Branch<K>
  fun search(k: K): K?
  fun traverse(f: (Node<K>) -> Unit): Unit
}

class Leaf<K: Comparable<K>>: Node<K>{
  override fun isRed(): Boolean = false
  override fun insert_(k: K): Branch<K>{
    return Branch(Color.Red, Leaf(), k, Leaf())
  }
  override fun search(k: K): K? = null
  override fun traverse(f: (Node<K>) -> Unit) = f(this)
}

class Branch<K: Comparable<K>>(val color: Color, val left: Node<K>, val key: K, val right: Node<K>): Node<K>{
  override fun isRed(): Boolean = color == Color.Red

  override fun insert_(k: K): Branch<K>{
    return if (k < key){
      balance(color, left.insert_(k), key, right)
    }else if (k > key){
      balance(color, left, key, right.insert_(k))
    }else{
      this
    }
  }

  override fun search(k: K): K? {
    return if (k < key){
      left.search(k)
    }else if (k > key){
      right.search(k)
    }else{
      key
    }
  }

  override fun traverse(f: (Node<K>) -> Unit){
    left.traverse(f)
    f(this)
    right.traverse(f)
  }

  private fun balance(c: Color, l: Node<K>, k: K, r: Node<K>): Branch<K>{
    if (c == Color.Red){
      return Branch(c, l, k, r)
    }

    if (l is Branch && l.isRed()){
      val ll = l.left
      val lr = l.right
      if (ll is Branch && ll.isRed()){
        return Branch(Color.Red,
        Branch(Color.Black, ll.left, ll.key, ll.right),
        l.key,
        Branch(Color.Black, lr, k, r))
      }
      else if (lr is Branch && lr.isRed()){
        return Branch(Color.Red,
        Branch(Color.Black, ll, l.key, lr.left),
        lr.key,
        Branch(Color.Black, lr.right, k, r))
      }
    }

    if (r is Branch && r.isRed()){
      val rl = r.left
      val rr = r.right
      if (rr is Branch && rr.isRed()){
        return Branch(Color.Red,
        Branch(Color.Black, l, k, rl),
        r.key,
        Branch(Color.Black, rr.left, rr.key, rr.right))
      }
      else if (rl is Branch && rl.isRed()){
        return Branch(Color.Red,
        Branch(Color.Black, l, k, rl.left),
        rl.key,
        Branch(Color.Black, rl.right, r.key, rr))
      }
    }

    return Branch(c, l, k, r)
  }
}

File: src/main/scala/RBTree.scala
/**
  * Created by chenl on 2016/11/17.
  */

import scala.math.Ordering.Implicits._

object RBTree {

  abstract class Color
  case class R() extends Color
  case class B() extends Color

  abstract class Tree[T: Ordering]()
  case class Node[T: Ordering](color: Color, left: Tree[T], key: T, right: Tree[T]) extends Tree[T]
  case class Empty[T: Ordering]() extends Tree[T]

  def insert[T: Ordering](e: T, tree: Tree[T]): Tree[T] = {
    def makeBlack(tree: Tree[T]): Tree[T] = tree match {
      case Node(_, l, k, r) => Node(B(), l, k, r)
    }

    def balance(color: Color, left: Tree[T], key: T, right: Tree[T]): Tree[T] = {
      (color, left, key, right) match {
        case (B(), Node(R(), Node(R(), a, x, b), y, c), z, d) => Node(R(), Node(B(), a, x, b), y, Node(B(), c, z, d))
        case (B(), Node(R(), a, x, Node(R(), b, y, c)), z, d) => Node(R(), Node(B(), a, x, b), y, Node(B(), c, z, d))
        case (B(), a, x, Node(R(), Node(R(), b, y, c), z, d)) => Node(R(), Node(B(), a, x, b), y, Node(B(), c, z, d))
        case (B(), a, x, Node(R(), b, y, Node(R(), c, z, d))) => Node(R(), Node(B(), a, x, b), y, Node(B(), c, z, d))
        case (c, l, x, r) => Node(c, l, x, r)
      }
    }

    def ins(tree: Tree[T]): Tree[T] = tree match {
        case Node(c, l, k, r) =>
          if (e < k)
            balance(c, ins(l), k, r)
          else if (e > k)
            balance(c, l, k, ins(r))
          else
            tree
        case Empty() => Node(R(), Empty(), e, Empty())
    }
    makeBlack(ins(tree))
  }
}
